p2p/simulations/mocker.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package simulations simulates p2p networks.
+// A mokcer simulates starting and stopping real nodes in a network.
+package simulations
+
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+)
+
+//a map of mocker names to its function
+var mockerList map[string]func(net *Network, quit chan struct{}, nodeCount int)
+
+//if the list has already been initialized
+var mockersInitialized bool
+
+//initialize the list of mockers
+func initializeMockers() {
+	//create the map
+	mockerList = make(map[string]func(net *Network, quit chan struct{}, nodeCount int))
+	//build the map
+	mockerList["startStop"] = startStop
+	mockerList["probabilistic"] = probabilistic
+	mockerList["boot"] = boot
+
+	mockersInitialized = true
+}
+
+//Lookup a mocker by its name, returns the mockerFn
+func LookupMocker(mockerType string) func(net *Network, quit chan struct{}, nodeCount int) {
+	if !mockersInitialized {
+		initializeMockers()
+	}
+	return mockerList[mockerType]
+}
+
+//Get a list of mockers (keys of the map)
+//Useful for frontend to build available mocker selection
+func GetMockerList() []string {
+	if !mockersInitialized {
+		initializeMockers()
+	}
+	list := make([]string, 0, len(mockerList))
+	for k, _ := range mockerList {
+		list = append(list, k)
+	}
+	return list
+}
+
+//The boot mockerFn only connects the node in a ring and doesn't do anything else
+func boot(net *Network, quit chan struct{}, nodeCount int) {
+	connectNodesInRing(net, nodeCount)
+}
+
+//The startStop mockerFn stops and starts nodes in a defined period (ticker)
+func startStop(net *Network, quit chan struct{}, nodeCount int) {
+	nodes := connectNodesInRing(net, nodeCount)
+	tick := time.NewTicker(10 * time.Second)
+	defer tick.Stop()
+	for {
+		select {
+		case <-quit:
+			log.Info("Terminating simulation loop")
+			return
+		case <-tick.C:
+			id := nodes[rand.Intn(len(nodes))]
+			go func() {
+				log.Info("stopping node", "id", id)
+				if err := net.Stop(id); err != nil {
+					log.Error("error stopping node", "id", id, "err", err)
+					return
+				}
+
+				time.Sleep(3 * time.Second)
```

Comment:
This should select on `quit` in case the mocking gets cancelled during this time.